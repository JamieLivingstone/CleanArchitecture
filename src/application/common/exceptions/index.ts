import { ZodError } from 'zod';

export class ValidationException extends Error {
  constructor(error: ZodError) {
    super('Validation failed');
    this.name = 'ValidationException';
    this.stack = new Error().stack;
    this.details = error.issues.map((issue) => ({
      path: issue.path.join('.'),
      message: issue.message,
    }));
  }

  details: { path: string; message: string }[];
}
