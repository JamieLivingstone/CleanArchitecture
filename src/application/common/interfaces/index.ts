export * from './logger';
export * from './posts-repository';
