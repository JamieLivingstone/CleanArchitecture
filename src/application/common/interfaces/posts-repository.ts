import { Post } from '@domain/entities';

export interface PostsRepository {
  create(parameters: { post: Post }): Promise<{ id: number }>;
  delete(parameters: { id: number }): Promise<void>;
  getById(parameters: { id: number }): Promise<Post | null>;
  list(parameters: { pageNumber: number; pageSize: number }): Promise<{ count: number; posts: Array<Post> }>;
  update(parameters: { post: Post }): Promise<void>;
}
