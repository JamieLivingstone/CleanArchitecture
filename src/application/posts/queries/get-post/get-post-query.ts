import { toDto } from './get-post-query-mapper';

export type GetPostQuery = Readonly<{
  id: number;
}>;

export function makeGetPostQuery({ postsRepository }: Pick<Dependencies, 'postsRepository'>) {
  return async function getPostQuery({ id }: GetPostQuery) {
    const post = await postsRepository.getById({ id });

    if (!post) {
      return null;
    }

    return toDto(post);
  };
}
