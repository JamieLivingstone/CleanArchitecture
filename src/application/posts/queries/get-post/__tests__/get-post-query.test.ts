import { PostsRepository } from '@application/common/interfaces';
import { Post } from '@domain/entities';
import { makeGetPostQuery } from '../get-post-query';

describe('getPostQuery', () => {
  function setup() {
    const postsRepository = jest.mocked<Partial<PostsRepository>>({
      getById: jest.fn(),
    }) as jest.Mocked<PostsRepository>;

    const getPostQuery = makeGetPostQuery({ postsRepository: postsRepository as PostsRepository });

    return {
      getPostQuery,
      postsRepository,
    };
  }

  describe('given the post does not exist', () => {
    it('should return null', async () => {
      // Arrange
      const { getPostQuery, postsRepository } = setup();
      postsRepository.getById.mockResolvedValue(null);

      // Act
      const result = await getPostQuery({ id: 1 });

      // Assert
      await expect(result).toEqual(null);
    });
  });

  describe('given the post exists', () => {
    it('should return the post', async () => {
      // Arrange
      const { getPostQuery, postsRepository } = setup();

      postsRepository.getById.mockResolvedValueOnce(
        new Post({
          id: 1,
          createdAt: new Date(2022, 1, 1),
          published: false,
          title: 'Mock post',
        }),
      );

      // Act
      const result = await getPostQuery({ id: 1 });

      // Assert
      expect(result).toMatchSnapshot();
    });
  });
});
