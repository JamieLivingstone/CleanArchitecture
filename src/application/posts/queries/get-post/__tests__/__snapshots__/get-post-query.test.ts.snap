// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPostQuery given the post exists should return the post 1`] = `
{
  "id": 1,
  "published": false,
  "title": "Mock post",
}
`;
